tycho-indexer/src/services/rpc.rs

```diff
@@ -718,6 +732,146 @@ where
+        let mut paginated_component_ids: Option<Vec<String>> = None;
+        if let Some(component_ids) = request.component_ids {
+            paginated_component_ids = Some(
+                component_ids
+                    .iter()
+                    .skip(pagination_params.offset() as usize)
+                    .take(pagination_params.page_size as usize)
+                    .cloned()
+                    .collect(),
+            );
+        }
+
+        let filter = EntryPointFilter {
+            protocol_system: request.protocol_system,
+            component_ids: paginated_component_ids,
+        };
+
+        let entry_points_tracing_params_data = self
+            .db_gateway
+            .get_entry_points_tracing_params(filter, Some(&pagination_params))
+            .await
+            .map_err(|err| {
+                error!(error = %err, "Error while getting entry points with tracing params.");
+                err
+            })?;
+
+        trace!(
+            entry_points_tracing_params = ?entry_points_tracing_params_data,
+            "Retrieved entry points with tracing params from database."
+        );
+
+        // Flatten the ID lists, throwing away component ids, to avoid making duplicate db calls
+        // when getting traced entry points.
+        let entry_point_ids: HashSet<EntryPointId> = entry_points_tracing_params_data
+            .entity
+            .values()
+            .flat_map(|entry_points_with_tracing_params| {
+                entry_points_with_tracing_params
+                    .iter()
+                    .map(|entry_point| {
+                        entry_point
+                            .entry_point
+                            .external_id
+                            .clone()
+                    })
+            })
+            .collect();
+
+        let traced_entry_points = self
+            .db_gateway
+            .get_traced_entry_points(&entry_point_ids)
+            .await
+            .map_err(|err| {
+                error!(error = %err, "Error while getting traced entry points.");
+                err
+            })?;
+
+        trace!(
+            traced_entry_points = ?traced_entry_points,
+            "Retrieved traced entry points entry points from database."
+        );
+
+        let mut traced_entry_points_by_component = HashMap::new();
+
+        for (component_id, entry_points_set) in entry_points_tracing_params_data.entity {
+            let mut pairs = Vec::new();
+
+            for entry_point_with_params in entry_points_set {
+                let entry_point_id = entry_point_with_params
+                    .entry_point
+                    .external_id
+                    .as_str()
+                    .to_string();
+                let tracing_param = &entry_point_with_params.params;
+
+                let results_by_param = traced_entry_points
+                    .get(&entry_point_id)
```

Comment:
FYI currently this still clones `entry_point_id` (the `to_string` methods is equivalent to a clone here), and then we only use the reference to `get()` in the hashmap. Instead we can completely avoid writing to the heap like this:

```suggestion
                let entry_point_id = entry_point_with_params
                    .entry_point
                    .external_id
                    .as_str();
                let tracing_param = &entry_point_with_params.params;

                let results_by_param = traced_entry_points
                    .get(entry_point_id)
```